sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryHashPartitionVerifySuite.scala

```diff
@@ -0,0 +1,199 @@
+
+  // Configs for random schema generation
+  private val variableTypes = Seq(IntegerType, DoubleType, FloatType, BinaryType, StringType)
+  val numSchemaTypes = 5
+  val maxNumFields = 10
+
+  // Configs for shuffle
+  val numRows = 10000
+  val numShufflePartitions = 100
+  private def saveSchemas(schemas: Seq[StructType]) = {
+    val writer = new BufferedWriter(new FileWriter(new File(goldenFileURI.getPath, schemaFileName)))
+    schemas.foreach{ schema =>
+      writer.write(schema.json)
+      writer.newLine()
+    }
+    writer.close()
+  }
+
+  private def readSchemas(): Seq[StructType] = {
+    val source = Source.fromFile(new File(goldenFileURI.getPath, schemaFileName))
+    try{
+      source.getLines().map{json => DataType.fromJson(json).asInstanceOf[StructType]}
+        .toArray // Avoid Stream lazy materialization
+        .toSeq
+    } finally
+      source.close()
+  }
+
+  private def saveRowsAndPartIds(rows: Seq[UnsafeRow], partIds: Seq[Int], os: DataOutputStream) = {
+    os.writeInt(rows.length)
+    rows.foreach { row =>
+      val rowBytes = row.getBytes()
+      os.writeInt(rowBytes.size)
+      os.write(rowBytes)}
+    partIds.foreach {id => os.writeInt(id)}
+  }
+
+  private def readRowsAndPartIds(is: DataInputStream): (Seq[UnsafeRow], Seq[Int]) = {
+    val numRows = is.readInt()
+    val rows = (1 to numRows).map{_ =>
+      val rowSize = is.readInt()
+      val rowBuffer = new Array[Byte](rowSize)
+      ByteStreams.readFully(is, rowBuffer, 0, rowSize)
+      val row = new UnsafeRow(1)
+      row.pointTo(rowBuffer, rowSize)
+      row
+    }
+    val partIds = (1 to numRows).map(_ => is.readInt()).toArray.toSeq
+    (rows, partIds)
+  }
+
+  private def getRandomRows(schema: StructType, numRows: Int, rand: Random): Seq[UnsafeRow] = {
+    val generator = RandomDataGenerator
+      .forType(schema, rand = new Random(rand.nextInt(RandomDataGenerator.MAX_STR_LEN)))
+      .get
+
+    // Create the converters needed to convert from external row to internal
+    // row and to UnsafeRows. Projection itself costs a lot on initialization
+    // (codegen and compile), so initialize it once.
+    val internalConverter = CatalystTypeConverters.createToCatalystConverter(schema)
+    val unsafeConverter = UnsafeProjection.create(Array(schema).asInstanceOf[Array[DataType]])
+
+    (1 to numRows).map { _ =>
+      val row = generator().asInstanceOf[Row]
+
+      val internalRow = new GenericInternalRow(1)
+      internalRow.update(0, internalConverter(row).asInstanceOf[GenericInternalRow])
+      val unsafeRow = unsafeConverter.apply(internalRow)
+      unsafeRow.copy()
+    }
+  }
+
+  private def getRandomSchemas(numSchemaTypes: Int, rand: Random,
+                               maxNumFields: Int): Seq[StructType] = {
+    (1 to numSchemaTypes).map { _ =>
+      val numField = rand.nextInt(maxNumFields)
+      RandomDataGenerator.randomSchema(rand, Math.max(1, numField), variableTypes)
```

Comment:
Should this be `Math.min`?

Reply:
This is to prevent the schema contains 0 field

Reply:
oh you are right..